pkg/kubectl/generate.go

```diff
@@ -0,0 +1,69 @@
+/*
+Copyright 2014 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubectl
+
+import (
+	"fmt"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/runtime"
+	"github.com/spf13/cobra"
+)
+
+// GeneratorParam is a parameter for a generator
+type GeneratorParam struct {
+	Name     string
+	Required bool
+}
+
+// Generator is an interface for things that can generate API objects from input parameters.
+type Generator interface {
+	// Generate creates an API object given a set of parameters
+	Generate(params map[string]string) (runtime.Object, error)
+	// ParamNames returns the list of parameters that this generator uses
+	ParamNames() []GeneratorParam
+}
+
+// Generators is a global list of known generators.
+// TODO: Dynamically create this from a list of template files?
+var Generators map[string]Generator = map[string]Generator{
+	"run-container-controller-v1": BasicReplicationController{},
```

Comment:
I would remove "controller" from the generator name, because:
1. it would more closely match the command name, thus more clearly establishing a pattern
2. it would reduce confusion about whether "run-container" is active in the same way a real API object is -- it's not, it's a macro expansion, which in this case just happens to expand into a controller
3. we're going to rename ReplicationController O(tomorrow)

I'd also like to use `/` as the separator between generator name and version, in case we want to reason about generator version in a more first-class manner: `run-container/v1`.